word_language_model/generate.py

```diff
@@ -49,19 +49,41 @@
 
 corpus = data.Corpus(args.data)
 ntokens = len(corpus.dictionary)
-hidden = model.init_hidden(1)
-input = torch.randint(ntokens, (1, 1), dtype=torch.long).to(device)
-
-with open(args.outf, 'w') as outf:
-    with torch.no_grad():  # no tracking history
-        for i in range(args.words):
-            output, hidden = model(input, hidden)
-            word_weights = output.squeeze().div(args.temperature).exp().cpu()
-            word_idx = torch.multinomial(word_weights, 1)[0]
-            input.fill_(word_idx)
-            word = corpus.dictionary.idx2word[word_idx]
-
-            outf.write(word + ('\n' if i % 20 == 19 else ' '))
-
-            if i % args.log_interval == 0:
-                print('| Generated {}/{} words'.format(i, args.words))
+
+if type(model).__name__ == "TransformerSeq2Seq":
+    input = torch.randint(ntokens, (1, 1), dtype=torch.long).to(device)
+
+    with open(args.outf, 'w') as outf:
+        with torch.no_grad():  # no tracking history
+            for i in range(args.words):
+                seq_mask = model.generate_square_subsequent_mask(input.size(0)).to(device)
+                output = model(input, input, src_mask=seq_mask, tgt_mask=seq_mask)
+                word_weights = output[-1].squeeze().div(args.temperature).exp().cpu()
```

Comment:
Why do we need to index into the last element of output and not do that same with the LSTM?

Reply:
AFAIK, in LSTM, the information from previous words are saving in the hidden and input is always a single word (here in this example). So the output from LSTM is also a single word. Differently in the transformer model, the information from previous words are carried by input. After each step, I attach the word_tensor to input by torch.cat. Therefore, the output is essentially a sentence. I dump the last word in the output to args.outf file.

Reply:
Is it ok to return the last word as the output or will it interfere with other use-sites?

Reply:
As in, the last word by default instead of the entire sentence.

Reply:
Let me run a test: feed last word to decoder and a sentence to encoder. If no difference, I will switch to the same setup as LSTM.